File: sorting/quicksort.php
<?php
/**
 * QuickSort Algorithm Implementation in PHP
 * 
 * Time Complexity:
 * - Best Case: O(n log n)
 * - Average Case: O(n log n)
 * - Worst Case: O(n²)
 * 
 * Space Complexity: O(log n) due to recursion stack
 * 
 * PHP features:
 * - Object-oriented design
 * - Type declarations
 * - Anonymous functions
 * - Array functions
 */

class QuickSort 
{
    /**
     * Sort array using QuickSort algorithm (returns new array)
     */
    public static function sort(array $array): array 
    {
        $result = $array;
        self::sortInPlace($result, 0, count($result) - 1);
        return $result;
    }
    
    /**
     * Sort array in-place using QuickSort algorithm
     */
    public static function sortInPlace(array &$array, int $low = null, int $high = null): void 
    {
        if ($low === null) $low = 0;
        if ($high === null) $high = count($array) - 1;
        
        if ($low < $high) {
            $pivotIndex = self::partition($array, $low, $high);
            self::sortInPlace($array, $low, $pivotIndex - 1);
            self::sortInPlace($array, $pivotIndex + 1, $high);
        }
    }
    
    /**
     * Partition the array around a pivot element
     */
    private static function partition(array &$array, int $low, int $high): int 
    {
        $pivot = $array[$high];
        $i = $low - 1;
        
        for ($j = $low; $j < $high; $j++) {
            if ($array[$j] <= $pivot) {
                $i++;
                self::swap($array, $i, $j);
            }
        }
        
        self::swap($array, $i + 1, $high);
        return $i + 1;
    }
    
    /**
     * Swap two elements in the array
     */
    private static function swap(array &$array, int $i, int $j): void 
    {
        if ($i !== $j) {
            $temp = $array[$i];
            $array[$i] = $array[$j];
            $array[$j] = $temp;
        }
    }
    
    /**
     * Iterative implementation to avoid stack overflow
     */
    public static function sortIterative(array &$array): void 
    {
        $size = count($array);
        if ($size <= 1) return;
        
        $stack = [];
        array_push($stack, [$low = 0, $high = $size - 1]);
        
        while (!empty($stack)) {
            [$low, $high] = array_pop($stack);
            
            if ($low < $high) {
                $pivotIndex = self::partition($array, $low, $high);
                array_push($stack, [$low, $pivotIndex - 1]);
                array_push($stack, [$pivotIndex + 1, $high]);
            }
        }
    }
    
    /**
     * Randomized QuickSort for better average performance
     */
    public static function sortRandomized(array &$array, int $low = null, int $high = null): void 
    {
        if ($low === null) $low = 0;
        if ($high === null) $high = count($array) - 1;
        
        if ($low < $high) {
            // Random pivot selection
            $randomIndex = rand($low, $high);
            self::swap($array, $randomIndex, $high);
            
            $pivotIndex = self::partition($array, $low, $high);
            self::sortRandomized($array, $low, $pivotIndex - 1);
            self::sortRandomized($array, $pivotIndex + 1, $high);
        }
    }
    
    /**
     * Three-way partitioning for arrays with many duplicates
     */
    public static function sort3Way(array &$array, int $low = null, int $high = null): void 
    {
        if ($low === null) $low = 0;
        if ($high === null) $high = count($array) - 1;
        
        if ($low >= $high) return;
        
        $pivot = $array[$low];
        $lt = $low;
        $gt = $high;
        $i = $low;
        
        while ($i <= $gt) {
            if ($array[$i] < $pivot) {
                self::swap($array, $lt, $i);
                $lt++;
                $i++;
            } elseif ($array[$i] > $pivot) {
                self::swap($array, $i, $gt);
                $gt--;
            } else {
                $i++;
            }
        }
        
        self::sort3Way($array, $low, $lt - 1);
        self::sort3Way($array, $gt + 1, $high);
    }
    
    /**
     * Functional-style QuickSort implementation
     */
    public static function sortFunctional(array $array): array 
    {
        if (count($array) <= 1) {
            return $array;
        }
        
        $pivot = $array[array_rand($array)];
        
        $less = array_filter($array, fn($x) => $x < $pivot);
        $equal = array_filter($array, fn($x) => $x == $pivot);
        $greater = array_filter($array, fn($x) => $x > $pivot);
        
        return array_merge(
            self::sortFunctional($less),
            array_values($equal),
            self::sortFunctional($greater)
        );
    }
    
    /**
     * Sort with custom comparison function
     */
    public static function sortWithComparator(array $array, callable $comparator): array 
    {
        $result = $array;
        self::sortWithComparatorInPlace($result, 0, count($result) - 1, $comparator);
        return $result;
    }
    
    private static function sortWithComparatorInPlace(array &$array, int $low, int $high, callable $comparator): void 
    {
        if ($low < $high) {
            $pivotIndex = self::partitionWithComparator($array, $low, $high, $comparator);
            self::sortWithComparatorInPlace($array, $low, $pivotIndex - 1, $comparator);
            self::sortWithComparatorInPlace($array, $pivotIndex + 1, $high, $comparator);
        }
    }
    
    private static function partitionWithComparator(array &$array, int $low, int $high, callable $comparator): int 
    {
        $pivot = $array[$high];
        $i = $low - 1;
        
        for ($j = $low; $j < $high; $j++) {
            if ($comparator($array[$j], $pivot) <= 0) {
                $i++;
                self::swap($array, $i, $j);
            }
        }
        
        self::swap($array, $i + 1, $high);
        return $i + 1;
    }
}

/**
 * Utility class for QuickSort operations
 */
class QuickSortUtils 
{
    /**
     * Check if array is sorted
     */
    public static function isSorted(array $array): bool 
    {
        $count = count($array);
        for ($i = 0; $i < $count - 1; $i++) {
            if ($array[$i] > $array[$i + 1]) {
                return false;
            }
        }
        return true;
    }
    
    /**
     * Generate random array for testing
     */
    public static function generateRandomArray(int $size, int $min = 1, int $max = 1000): array 
    {
        $array = [];
        for ($i = 0; $i < $size; $i++) {
            $array[] = rand($min, $max);
        }
        return $array;
    }
    
    /**
     * Benchmark sorting performance
     */
    public static function benchmark(array $array, string $method = 'sort'): float 
    {
        $testArray = $array;
        $startTime = microtime(true);
        
        switch ($method) {
            case 'sort':
                QuickSort::sort($testArray);
                break;
            case 'iterative':
                QuickSort::sortIterative($testArray);
                break;
            case 'randomized':
                QuickSort::sortRandomized($testArray);
                break;
            case 'functional':
                QuickSort::sortFunctional($testArray);
                break;
            case '3way':
                QuickSort::sort3Way($testArray);
                break;
            default:
                throw new InvalidArgumentException("Unknown method: $method");
        }
        
        return (microtime(true) - $startTime) * 1000; // Return milliseconds
    }
    
    /**
     * Pretty print array
     */
    public static function printArray(array $array, string $label = ''): void 
    {
        if ($label) {
            echo "$label: ";
        }
        echo '[' . implode(', ', $array) . ']' . PHP_EOL;
    }
}

/**
 * Array extension trait (simulating extension methods)
 */
trait QuickSortArrayExtensions 
{
    public function quickSort(): array 
    {
        return QuickSort::sort($this);
    }
    
    public function quickSortInPlace(): void 
    {
        QuickSort::sortInPlace($this);
    }
    
    public function isSorted(): bool 
    {
        return QuickSortUtils::isSorted($this);
    }
}

/**
 * Enhanced array class with QuickSort methods
 */
class SortableArray extends ArrayObject 
{
    use QuickSortArrayExtensions;
    
    public function quickSort(): SortableArray 
    {
        $sorted = QuickSort::sort($this->getArrayCopy());
        return new self($sorted);
    }
    
    public function quickSortInPlace(): self 
    {
        $array = $this->getArrayCopy();
        QuickSort::sortInPlace($array);
        $this->exchangeArray($array);
        return $this;
    }
}

/**
 * Demonstration and testing function
 */
function demonstrateQuickSort(): void 
{
    echo "🚀 PHP QuickSort Implementation" . PHP_EOL;
    echo str_repeat('=', 40) . PHP_EOL;
    
    // Test arrays
    $testArrays = [
        [64, 34, 25, 12, 22, 11, 90],
        [5, 2, 8, 6, 1, 9, 4],
        [1],
        [],
        [3, 3, 3, 3, 3],
        [9, 8, 7, 6, 5, 4, 3, 2, 1],
    ];
    
    echo PHP_EOL . "📋 Basic Sorting Tests:" . PHP_EOL;
    echo str_repeat('-', 30) . PHP_EOL;
    
    foreach ($testArrays as $index => $array) {
        $original = $array;
        $sorted = QuickSort::sort($array);
        $isCorrect = QuickSortUtils::isSorted($sorted);
        
        echo "Test " . ($index + 1) . ":" . PHP_EOL;
        QuickSortUtils::printArray($original, "Original");
        QuickSortUtils::printArray($sorted, "Sorted  ");
        echo "Correct: " . ($isCorrect ? "✓" : "✗") . PHP_EOL;
        echo str_repeat("-", 25) . PHP_EOL;
    }
    
    // Advanced features demonstration
    echo PHP_EOL . "🎯 Advanced Features:" . PHP_EOL;
    echo str_repeat('-', 25) . PHP_EOL;
    
    // Custom comparator (reverse order)
    $numbers = [64, 34, 25, 12, 22, 11, 90];
    $reverseSorted = QuickSort::sortWithComparator(
        $numbers, 
        fn($a, $b) => $b <=> $a
    );
    
    QuickSortUtils::printArray($numbers, "Original");
    QuickSortUtils::printArray($reverseSorted, "Reverse ");
    
    // Functional approach
    $functionalSorted = QuickSort::sortFunctional($numbers);
    QuickSortUtils::printArray($functionalSorted, "Functional");
    
    // Using enhanced array class
    $sortableArray = new SortableArray([9, 3, 7, 1, 5]);
    echo PHP_EOL . "Enhanced Array Class:" . PHP_EOL;
    QuickSortUtils::printArray($sortableArray->getArrayCopy(), "Original");
    $sortableArray->quickSortInPlace();
    QuickSortUtils::printArray($sortableArray->getArrayCopy(), "Sorted  ");
    
    // String sorting
    $words = ['banana', 'apple', 'cherry', 'date'];
    $sortedWords = QuickSort::sort($words);
    QuickSortUtils::printArray($words, "Words Original");
    QuickSortUtils::printArray($sortedWords, "Words Sorted  ");
}

/**
 * Performance benchmarking function
 */
function performanceBenchmark(): void 
{
    echo PHP_EOL . "⚡ Performance Benchmark" . PHP_EOL;
    echo str_repeat('=', 30) . PHP_EOL;
    
    $sizes = [1000, 5000, 10000];
    
    foreach ($sizes as $size) {
        echo PHP_EOL . "Testing with $size elements:" . PHP_EOL;
        
        $testData = QuickSortUtils::generateRandomArray($size);
        $methods = ['sort', 'iterative', 'randomized', '3way', 'functional'];
        
        foreach ($methods as $method) {
            $time = QuickSortUtils::benchmark($testData, $method);
            printf("  %-12s: %8.2fms%s", 
                ucfirst($method), 
                $time, 
                PHP_EOL
            );
        }
    }
}

/**
 * Edge cases testing
 */
function testEdgeCases(): void 
{
    echo PHP_EOL . "🧩 Edge Cases Testing" . PHP_EOL;
    echo str_repeat('=', 25) . PHP_EOL;
    
    // Test edge cases
    $edgeCases = [
        'Empty array' => [],
        'Single element' => [42],
        'Two elements' => [2, 1],
        'Already sorted' => [1, 2, 3, 4, 5],
        'Reverse sorted' => [5, 4, 3, 2, 1],
        'All duplicates' => [5, 5, 5, 5, 5],
    ];
    
    foreach ($edgeCases as $name => $array) {
        $sorted = QuickSort::sort($array);
        $isCorrect = QuickSortUtils::isSorted($sorted);
        echo "$name: " . ($isCorrect ? "✓" : "✗") . PHP_EOL;
    }
}

// Main execution
if (php_sapi_name() === 'cli') {
    demonstrateQuickSort();
    performanceBenchmark();
    testEdgeCases();
    
    echo PHP_EOL . "✨ QuickSort demonstration complete!" . PHP_EOL;
}
?>

File: sorting/quicksort.kt
/**
 * QuickSort Algorithm Implementation in Kotlin
 * 
 * Time Complexity:
 * - Best Case: O(n log n)
 * - Average Case: O(n log n)
 * - Worst Case: O(n²)
 * 
 * Space Complexity: O(log n) due to recursion stack
 * 
 * Kotlin features:
 * - Extension functions
 * - Higher-order functions
 * - Null safety
 * - Data classes and sealed classes
 * - Coroutines support
 */

import kotlin.random.Random
import kotlin.system.measureTimeMillis
import kotlinx.coroutines.*

/**
 * QuickSort implementation with multiple variants
 */
class QuickSort {
    companion object {
        /**
         * Sort array using QuickSort algorithm (returns new array)
         */
        fun <T : Comparable<T>> sort(array: Array<T>): Array<T> {
            val result = array.copyOf()
            sortInPlace(result, 0, result.size - 1)
            return result
        }
        
        /**
         * Sort array in-place using QuickSort algorithm
         */
        fun <T : Comparable<T>> sortInPlace(array: Array<T>, low: Int = 0, high: Int = array.size - 1) {
            if (low < high) {
                val pivotIndex = partition(array, low, high)
                sortInPlace(array, low, pivotIndex - 1)
                sortInPlace(array, pivotIndex + 1, high)
            }
        }
        
        /**
         * Partition the array around a pivot element
         */
        private fun <T : Comparable<T>> partition(array: Array<T>, low: Int, high: Int): Int {
            val pivot = array[high]
            var i = low - 1
            
            for (j in low until high) {
                if (array[j] <= pivot) {
                    i++
                    array.swap(i, j)
                }
            }
            
            array.swap(i + 1, high)
            return i + 1
        }
        
        /**
         * Iterative implementation to avoid stack overflow
         */
        fun <T : Comparable<T>> sortIterative(array: Array<T>) {
            if (array.size <= 1) return
            
            val stack = mutableListOf<Pair<Int, Int>>()
            stack.add(0 to array.size - 1)
            
            while (stack.isNotEmpty()) {
                val (low, high) = stack.removeAt(stack.size - 1)
                
                if (low < high) {
                    val pivotIndex = partition(array, low, high)
                    stack.add(low to pivotIndex - 1)
                    stack.add(pivotIndex + 1 to high)
                }
            }
        }
        
        /**
         * Randomized QuickSort for better average performance
         */
        fun <T : Comparable<T>> sortRandomized(array: Array<T>, low: Int = 0, high: Int = array.size - 1) {
            if (low < high) {
                // Random pivot selection
                val randomIndex = Random.nextInt(low, high + 1)
                array.swap(randomIndex, high)
                
                val pivotIndex = partition(array, low, high)
                sortRandomized(array, low, pivotIndex - 1)
                sortRandomized(array, pivotIndex + 1, high)
            }
        }
        
        /**
         * Three-way partitioning for arrays with many duplicates
         */
        fun <T : Comparable<T>> sort3Way(array: Array<T>, low: Int = 0, high: Int = array.size - 1) {
            if (low >= high) return
            
            val pivot = array[low]
            var lt = low
            var gt = high
            var i = low
            
            while (i <= gt) {
                when {
                    array[i] < pivot -> {
                        array.swap(lt, i)
                        lt++
                        i++
                    }
                    array[i] > pivot -> {
                        array.swap(i, gt)
                        gt--
                    }
                    else -> i++
                }
            }
            
            sort3Way(array, low, lt - 1)
            sort3Way(array, gt + 1, high)
        }
        
        /**
         * Functional-style QuickSort implementation
         */
        fun <T : Comparable<T>> sortFunctional(list: List<T>): List<T> {
            if (list.size <= 1) return list
            
            val pivot = list[list.size / 2]
            val less = list.filter { it < pivot }
            val equal = list.filter { it == pivot }
            val greater = list.filter { it > pivot }
            
            return sortFunctional(less) + equal + sortFunctional(greater)
        }
        
        /**
         * Sort with custom comparator
         */
        fun <T> sortWithComparator(array: Array<T>, comparator: Comparator<T>): Array<T> {
            val result = array.copyOf()
            sortWithComparatorInPlace(result, 0, result.size - 1, comparator)
            return result
        }
        
        private fun <T> sortWithComparatorInPlace(
            array: Array<T>, 
            low: Int, 
            high: Int, 
            comparator: Comparator<T>
        ) {
            if (low < high) {
                val pivotIndex = partitionWithComparator(array, low, high, comparator)
                sortWithComparatorInPlace(array, low, pivotIndex - 1, comparator)
                sortWithComparatorInPlace(array, pivotIndex + 1, high, comparator)
            }
        }
        
        private fun <T> partitionWithComparator(
            array: Array<T>, 
            low: Int, 
            high: Int, 
            comparator: Comparator<T>
        ): Int {
            val pivot = array[high]
            var i = low - 1
            
            for (j in low until high) {
                if (comparator.compare(array[j], pivot) <= 0) {
                    i++
                    array.swap(i, j)
                }
            }
            
            array.swap(i + 1, high)
            return i + 1
        }
        
        /**
         * Parallel QuickSort using coroutines
         */
        suspend fun <T : Comparable<T>> sortParallel(
            array: Array<T>, 
            low: Int = 0, 
            high: Int = array.size - 1,
            threshold: Int = 1000
        ) {
            if (low < high) {
                val pivotIndex = partition(array, low, high)
                
                if (high - low > threshold) {
                    // Use parallel processing for large subarrays
                    coroutineScope {
                        launch { sortParallel(array, low, pivotIndex - 1, threshold) }
                        launch { sortParallel(array, pivotIndex + 1, high, threshold) }
                    }
                } else {
                    // Use sequential processing for small subarrays
                    sortInPlace(array, low, pivotIndex - 1)
                    sortInPlace(array, pivotIndex + 1, high)
                }
            }
        }
    }
}

/**
 * Extension functions for arrays and lists
 */
fun <T : Comparable<T>> Array<T>.quickSort(): Array<T> = QuickSort.sort(this)

fun <T : Comparable<T>> Array<T>.quickSortInPlace() = QuickSort.sortInPlace(this)

fun <T : Comparable<T>> List<T>.quickSort(): List<T> = QuickSort.sortFunctional(this)

fun <T : Comparable<T>> Array<T>.isSorted(): Boolean {
    for (i in 0 until size - 1) {
        if (this[i] > this[i + 1]) return false
    }
    return true
}

fun <T> Array<T>.swap(i: Int, j: Int) {
    if (i != j) {
        val temp = this[i]
        this[i] = this[j]
        this[j] = temp
    }
}

/**
 * Utility class for QuickSort operations
 */
object QuickSortUtils {
    /**
     * Generate random array for testing
     */
    fun generateRandomArray(size: Int, range: IntRange = 1..1000): Array<Int> {
        return Array(size) { Random.nextInt(range.first, range.last + 1) }
    }
    
    /**
     * Benchmark sorting performance
     */
    fun <T : Comparable<T>> benchmark(
        array: Array<T>, 
        method: String = "standard"
    ): Long {
        val testArray = array.copyOf()
        
        return measureTimeMillis {
            when (method.lowercase()) {
                "standard" -> QuickSort.sortInPlace(testArray)
                "iterative" -> QuickSort.sortIterative(testArray)
                "randomized" -> QuickSort.sortRandomized(testArray)
                "3way" -> QuickSort.sort3Way(testArray)
                else -> throw IllegalArgumentException("Unknown method: $method")
            }
        }
    }
    
    /**
     * Pretty print array
     */
    fun <T> printArray(array: Array<T>, label: String = "") {
        if (label.isNotEmpty()) {
            print("$label: ")
        }
        println(array.contentToString())
    }
}

/**
 * Data class for performance results
 */
data class BenchmarkResult(
    val method: String,
    val arraySize: Int,
    val timeMs: Long,
    val isCorrect: Boolean
)

/**
 * Sealed class for different sorting strategies
 */
sealed class SortingStrategy {
    object Standard : SortingStrategy()
    object Iterative : SortingStrategy()
    object Randomized : SortingStrategy()
    object ThreeWay : SortingStrategy()
    object Functional : SortingStrategy()
    
    fun <T : Comparable<T>> sort(array: Array<T>): Array<T> {
        return when (this) {
            is Standard -> QuickSort.sort(array)
            is Iterative -> array.copyOf().also { QuickSort.sortIterative(it) }
            is Randomized -> array.copyOf().also { QuickSort.sortRandomized(it) }
            is ThreeWay -> array.copyOf().also { QuickSort.sort3Way(it) }
            is Functional -> QuickSort.sortFunctional(array.toList()).toTypedArray()
        }
    }
    
    override fun toString(): String = this::class.simpleName ?: "Unknown"
}

/**
 * QuickSort analyzer class
 */
class QuickSortAnalyzer {
    fun analyzePerformance(sizes: List<Int>, iterations: Int = 5): List<BenchmarkResult> {
        val results = mutableListOf<BenchmarkResult>()
        val strategies = listOf(
            SortingStrategy.Standard,
            SortingStrategy.Iterative,
            SortingStrategy.Randomized,
            SortingStrategy.ThreeWay
        )
        
        for (size in sizes) {
            println("Testing with $size elements:")
            
            for (strategy in strategies) {
                val times = mutableListOf<Long>()
                var isCorrect = true
                
                repeat(iterations) {
                    val testArray = QuickSortUtils.generateRandomArray(size)
                    val originalArray = testArray.copyOf()
                    
                    val time = measureTimeMillis {
                        val sorted = strategy.sort(testArray)
                        if (!sorted.isSorted()) isCorrect = false
                    }
                    times.add(time)
                }
                
                val avgTime = times.average().toLong()
                results.add(BenchmarkResult(strategy.toString(), size, avgTime, isCorrect))
                
                println("  ${strategy.toString().padEnd(12)}: ${avgTime}ms avg ${if (isCorrect) "✓" else "✗"}")
            }
            println()
        }
        
        return results
    }
}

/**
 * Demonstration and testing functions
 */
fun demonstrateQuickSort() {
    println("🚀 Kotlin QuickSort Implementation")
    println("=".repeat(40))
    
    // Test arrays
    val testArrays = listOf(
        arrayOf(64, 34, 25, 12, 22, 11, 90),
        arrayOf(5, 2, 8, 6, 1, 9, 4),
        arrayOf(1),
        arrayOf<Int>(),
        arrayOf(3, 3, 3, 3, 3),
        arrayOf(9, 8, 7, 6, 5, 4, 3, 2, 1)
    )
    
    println("\n📋 Basic Sorting Tests:")
    println("-".repeat(30))
    
    testArrays.forEachIndexed { index, array ->
        val original = array.copyOf()
        val sorted = array.quickSort()
        val isCorrect = sorted.isSorted()
        
        println("Test ${index + 1}:")
        QuickSortUtils.printArray(original, "Original")
        QuickSortUtils.printArray(sorted, "Sorted  ")
        println("Correct: ${if (isCorrect) "✓" else "✗"}")
        println("-".repeat(25))
    }
    
    // Advanced features demonstration
    println("\n🎯 Advanced Features:")
    println("-".repeat(25))
    
    // String sorting
    val words = arrayOf("banana", "apple", "cherry", "date", "elderberry")
    val sortedWords = words.quickSort()
    QuickSortUtils.printArray(words, "Original words")
    QuickSortUtils.printArray(sortedWords, "Sorted words  ")
    
    // Custom comparator (reverse order)
    val numbers = arrayOf(64, 34, 25, 12, 22, 11, 90)
    val reverseSorted = QuickSort.sortWithComparator(numbers) { a, b -> b.compareTo(a) }
    QuickSortUtils.printArray(numbers, "Original  ")
    QuickSortUtils.printArray(reverseSorted, "Reverse   ")
    
    // Functional approach with lists
    val numberList = listOf(9, 3, 7, 1, 5)
    val functionalSorted = numberList.quickSort()
    println("Original list:   $numberList")
    println("Functional sort: $functionalSorted")
    
    // Using extension functions
    val extArray = arrayOf(42, 17, 89, 3, 56)
    println("\nExtension functions:")
    QuickSortUtils.printArray(extArray, "Before")
    extArray.quickSortInPlace()
    QuickSortUtils.printArray(extArray, "After ")
}

suspend fun demonstrateParallelSort() {
    println("\n🔄 Parallel Sorting Demonstration:")
    println("-".repeat(35))
    
    val largeArray = QuickSortUtils.generateRandomArray(100000)
    val testArray1 = largeArray.copyOf()
    val testArray2 = largeArray.copyOf()
    
    // Sequential sort
    val sequentialTime = measureTimeMillis {
        QuickSort.sortInPlace(testArray1)
    }
    
    // Parallel sort
    val parallelTime = measureTimeMillis {
        runBlocking {
            QuickSort.sortParallel(testArray2)
        }
    }
    
    println("Array size: ${largeArray.size}")
    println("Sequential time: ${sequentialTime}ms")
    println("Parallel time:   ${parallelTime}ms")
    println("Speedup: ${sequentialTime.toDouble() / parallelTime}x")
    println("Both correct: ${testArray1.isSorted() && testArray2.isSorted()}")
}

fun performanceBenchmark() {
    println("\n⚡ Performance Benchmark")
    println("=".repeat(30))
    
    val analyzer = QuickSortAnalyzer()
    val sizes = listOf(1000, 5000, 10000)
    analyzer.analyzePerformance(sizes, iterations = 3)
}

fun testEdgeCases() {
    println("🧩 Edge Cases Testing")
    println("=".repeat(25))
    
    val edgeCases = mapOf(
        "Empty array" to arrayOf<Int>(),
        "Single element" to arrayOf(42),
        "Two elements" to arrayOf(2, 1),
        "Already sorted" to arrayOf(1, 2, 3, 4, 5),
        "Reverse sorted" to arrayOf(5, 4, 3, 2, 1),
        "All duplicates" to arrayOf(5, 5, 5, 5, 5)
    )
    
    edgeCases.forEach { (name, array) ->
        val sorted = array.quickSort()
        val isCorrect = sorted.isSorted()
        println("$name: ${if (isCorrect) "✓" else "✗"}")
    }
}

/**
 * Main function
 */
suspend fun main() {
    demonstrateQuickSort()
    demonstrateParallelSort()
    performanceBenchmark()
    testEdgeCases()
    
    println("\n✨ QuickSort demonstration complete!")
}
